src/main/scala/cognite/spark/v1/RawTableRelation.scala

```diff
@@ -46,28 +46,36 @@ class RawTableRelation(
 
   override val schema: StructType = userSchema.getOrElse {
     if (inferSchema) {
-      val rdd =
-        readRows(
-          inferSchemaLimit.orElse(Some(Constants.DefaultInferSchemaLimit)),
-          Some(1),
-          RawRowFilter(),
-          None,
-          collectSchemaInferenceMetrics,
-          false)
-
-      import sqlContext.sparkSession.implicits._
-      val df = sqlContext.createDataFrame(rdd, defaultSchema)
-      val jsonDf =
-        renameColumns(sqlContext.sparkSession.read.json(df.select($"columns").as[String]))
+      val jsonSchema =
+        inferSchema(inferSchemaLimit.getOrElse(Constants.DefaultInferSchemaLimit))
+          .unsafeRunSync()
+      val renamedJsonSchema = schemaWithRenamedColumns(jsonSchema)
       StructType(
         StructField("key", DataTypes.StringType, nullable = false)
           +: StructField(lastUpdatedTimeColName, DataTypes.TimestampType, nullable = true)
-          +: jsonDf.schema.fields)
+          +: renamedJsonSchema.fields)
     } else {
       defaultSchema
     }
   }
 
+  private def inferSchema(
+      limit: Int
+  ): IO[StructType] = {
+    val client = CdpConnector.clientFromConfig(config).rawRows(database, table)
+
+    val rows = client
+      .list(Some(limit))
```

Comment:
This isn't the one you want, and probably why it's not so fast, as it's not doing concurrent reads. ;) You should use `listConcurrently`. If you want to limit, just limit the resulting (combined) `fs2.Stream`.

More like we're doing here: https://github.com/cognitedata/profiler-api/blob/main/src/main/scala/com/cognite/raw_profiler/raw/RawClient.scala#L116

fwiw `Stream(streams: _*).parJoin(10)` seems to be a bit faster than the `.fold(Stream.empty)(_.merge(_))` we're doing in the Scala SDK.

Reply:
will do that :+1:. Currently it's not an issue, since we still use the 10k (one page) limit, but it will be very important for increasing the infer schema limit.